### DISCUSSION TITLE
[Orange Map] How should I type values which are `BigInt`?

### DISCUSSION THREAD

**Discussion by Zer0xxxx** [alfateam/rdb#101] [Q&A] [answered]
At first, I thought that a `BIGINT`, or `BIGSERIAL` should be mapped using `numeric()`.

In practice, querying the DB like so:
```ts
const guild = await db.guild.insert({
  discord_guild_id: BigInt('123456789012345678901234567890'),
  lang: 'en-US'
});
```

Where `discord_guild_id` is mapped with `numeric()`,

Causes a type error:
> No overload matches this call.

I _think_ this issue is related to this TypeScript subtlety:
> Type `bigint` is not assignable to type `number`

So...
Should I map it as a `string()` and validate it via a RegEx as a workaround?
Then I could retrieve it via the ORM, and pass it back into `BigInt` to get the correct value?

(Btw: this is what other Node.js ORMs suggest to do... even if I'm unhappy with this. :/)

---

Btw, the JavaScript `Number` constructor is unsafe when handling big numbers:

![image](https://github.com/alfateam/orange-orm/assets/36828641/98f7204e-b9e9-4c1d-ab9c-79c11bb11e4a)

**Comment by lroal** [comment]
If the bigint value from the db exceeds 9007199254740991, you need to map it as string. 
But you would safe yourself a lot of trouble staying in safe integer zone. You could have your valid use case, but why do you want bigint ?

**Reply by Zer0xxxx** [reply]
> If the bigint value from the db exceeds 9007199254740991, you need to map it as string. But you would safe yourself a lot of trouble staying in safe integer zone. You could have your valid use case, but why do you want bigint ?

It does not depend on me.
Here, I'm depending on a third-party API, which is the Discord API.
All the IDs provided by this API are `BigInt` (channels, guilds, users... every "ID").

**Reply by lroal** [reply]
If the string workaround does not work or is cumbersome, I can always create an explicit mapping for bigint

**Reply by Zer0xxxx** [reply]
As Orange ORM is greatly typed, and since I liked it because of this philosophy... I think it would be awesome!
Orange ORM is an extremely well-defined solution, well mathematically proofed, with which everything makes sense and becomes as easy as it should be.
I don't feel comfortable starting to make "Workarounds" with this lib, especially for primitives such as BigInt. :(

**Reply by Zer0xxxx** [reply]
@lroal Do you think I should create an issue for this?

**Reply by lroal** [reply]
Yes

**Comment by lroal** [comment, marked as answer]
Done in version 4.7.11
